File: examples/mcp_client_example.js
/**
 * Simple MCP client example for testing the MasterControl Android app.
 * This script demonstrates how to connect to the MCP server and call tools.
 * 
 * Requirements:
 *     npm install ws
 * 
 * Usage:
 *     node mcp_client_example.js
 */

const WebSocket = require('ws');

function sendRequest(ws, method, params = null, requestId = 1) {
    return new Promise((resolve, reject) => {
        const request = {
            jsonrpc: "2.0",
            id: requestId,
            method: method
        };
        
        if (params) {
            request.params = params;
        }
        
        console.log(`\n→ Sending: ${JSON.stringify(request, null, 2)}`);
        
        // Set up one-time listener for response
        const responseHandler = (data) => {
            console.log(`← Received: ${data}`);
            ws.removeListener('message', responseHandler);
            resolve(JSON.parse(data));
        };
        
        ws.on('message', responseHandler);
        ws.send(JSON.stringify(request));
    });
}

async function main() {
    console.log("Connecting to MCP server at ws://localhost:8080/mcp...");
    
    const ws = new WebSocket('ws://localhost:8080/mcp');
    
    await new Promise((resolve, reject) => {
        ws.on('open', resolve);
        ws.on('error', reject);
    });
    
    console.log("✓ Connected!");
    
    try {
        // 1. Initialize
        console.log("\n=== 1. Initialize ===");
        let response = await sendRequest(ws, "initialize", null, 1);
        console.log(`Server: ${response.result.serverInfo.name} v${response.result.serverInfo.version}`);
        
        // 2. List tools
        console.log("\n=== 2. List Available Tools ===");
        response = await sendRequest(ws, "tools/list", null, 2);
        const tools = response.result.tools;
        console.log(`Found ${tools.length} tools:`);
        tools.forEach(tool => {
            console.log(`  - ${tool.name}: ${tool.description}`);
        });
        
        // 3. Call echo tool
        console.log("\n=== 3. Call Echo Tool ===");
        response = await sendRequest(ws, "tools/call", {
            name: "echo",
            arguments: {
                message: "Hello from Node.js client!"
            }
        }, 3);
        const echoResult = response.result.content[0].text;
        console.log(`Echo result: ${echoResult}`);
        
        // 4. Get device info
        console.log("\n=== 4. Get Device Info ===");
        response = await sendRequest(ws, "tools/call", {
            name: "get_device_info",
            arguments: {}
        }, 4);
        const deviceInfo = response.result.content[0].text;
        console.log(`Device info:\n${deviceInfo}`);
        
        // 5. Send notification
        console.log("\n=== 5. Send Notification ===");
        response = await sendRequest(ws, "tools/call", {
            name: "send_notification",
            arguments: {
                title: "MCP Test",
                message: "Notification sent from Node.js client!"
            }
        }, 5);
        const notifResult = response.result.content[0].text;
        console.log(`Notification result: ${notifResult}`);
        
        console.log("\n✓ All tests completed successfully!");
        
    } catch (error) {
        console.error(`\n✗ Error: ${error.message}`);
    } finally {
        ws.close();
        console.log("\n✓ Connection closed");
    }
}

main().catch(error => {
    console.error(`\n✗ Error: ${error.message}`);
    console.log("\nMake sure:");
    console.log("  1. The MasterControl app is running on your device");
    console.log("  2. The MCP server is started (tap 'Start MCP Server' button)");
    console.log("  3. You're connected to the same network or using adb port forwarding:");
    console.log("     adb forward tcp:8080 tcp:8080");
    process.exit(1);
});

File: examples/mcp_client_example.py
#!/usr/bin/env python3
"""
Simple MCP client example for testing the MasterControl Android app.
This script demonstrates how to connect to the MCP server and call tools.

Requirements:
    pip install websocket-client

Usage:
    python mcp_client_example.py
"""

import json
import websocket

def send_request(ws, method, params=None, request_id=1):
    """Send an MCP request and return the response."""
    request = {
        "jsonrpc": "2.0",
        "id": request_id,
        "method": method
    }
    if params:
        request["params"] = params
    
    print(f"\n→ Sending: {json.dumps(request, indent=2)}")
    ws.send(json.dumps(request))
    
    response = ws.recv()
    print(f"← Received: {response}")
    return json.loads(response)

def main():
    # Connect to MCP server
    print("Connecting to MCP server at ws://localhost:8080/mcp...")
    ws = websocket.create_connection("ws://localhost:8080/mcp")
    print("✓ Connected!")
    
    try:
        # 1. Initialize
        print("\n=== 1. Initialize ===")
        response = send_request(ws, "initialize", request_id=1)
        print(f"Server: {response['result']['serverInfo']['name']} v{response['result']['serverInfo']['version']}")
        
        # 2. List tools
        print("\n=== 2. List Available Tools ===")
        response = send_request(ws, "tools/list", request_id=2)
        tools = response['result']['tools']
        print(f"Found {len(tools)} tools:")
        for tool in tools:
            print(f"  - {tool['name']}: {tool['description']}")
        
        # 3. Call echo tool
        print("\n=== 3. Call Echo Tool ===")
        response = send_request(
            ws,
            "tools/call",
            params={
                "name": "echo",
                "arguments": {
                    "message": "Hello from Python client!"
                }
            },
            request_id=3
        )
        result = response['result']['content'][0]['text']
        print(f"Echo result: {result}")
        
        # 4. Get device info
        print("\n=== 4. Get Device Info ===")
        response = send_request(
            ws,
            "tools/call",
            params={
                "name": "get_device_info",
                "arguments": {}
            },
            request_id=4
        )
        device_info = response['result']['content'][0]['text']
        print(f"Device info:\n{device_info}")
        
        # 5. Send notification
        print("\n=== 5. Send Notification ===")
        response = send_request(
            ws,
            "tools/call",
            params={
                "name": "send_notification",
                "arguments": {
                    "title": "MCP Test",
                    "message": "Notification sent from Python client!"
                }
            },
            request_id=5
        )
        result = response['result']['content'][0]['text']
        print(f"Notification result: {result}")
        
        print("\n✓ All tests completed successfully!")
        
    finally:
        ws.close()
        print("\n✓ Connection closed")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"\n✗ Error: {e}")
        print("\nMake sure:")
        print("  1. The MasterControl app is running on your device")
        print("  2. The MCP server is started (tap 'Start MCP Server' button)")
        print("  3. You're connected to the same network or using adb port forwarding:")
        print("     adb forward tcp:8080 tcp:8080")
